lib/dynamic_supervisor.ex

```diff
@@ -427,9 +440,85 @@ defmodule DynamicSupervisor do
+  defp validate_dynamic(:infinity), do: :ok
+  defp validate_dynamic(dynamic) when is_integer(dynamic), do: :ok
+  defp validate_dynamic(_), do: {:error, "max_dynamic must be an integer or :infinity"}
+
+  @doc false
+  def handle_subscribe(opts, {_, ref} = from, state) do
+    ## GenStage checks these options before allowing susbcription
+    max = Keyword.get(opts, :max_demand, 100)
+    min = Keyword.get(opts, :min_demand, div(max, 2))
+    GenStage.ask(from, max)
+    {:manual, put_in(state.producers[ref], {from, 0, max, min, max})}
+  end
+
+  @doc false
+  def handle_cancel(_, {_, ref}, state) do
+    {:noreply, update_in(state.producers, &Map.delete(&1, ref))}
+  end
+
+  @doc false
+  def handle_events(events, {_, ref} = from, state) do
+    %{template: child, children: children} = state
+    {new, errors} = start_children(events, from, child, 0, [], state)
+    new_children = Enum.into(new, children)
+    started = map_size(new_children) - map_size(children)
+    {:noreply, [], maybe_ask(ref, started+errors, errors, new_children, state)}
+  end
+
+  defp start_children([extra | extras], from, child, errors, acc, state) do
+    {_, ref} = from
+    {_, {m, f, args}, restart, _, _, _} = child
+    args = args ++ extra
+    case start_child(m, f, args) do
+      {:ok, pid, _} when restart == :temporary ->
+        acc = [{pid, [ref | :undefined]} | acc]
+        start_children(extras, from, child, errors, acc, state)
+      {:ok, pid, _}  ->
+        acc = [{pid, [ref | args]} | acc]
+        start_children(extras, from, child, errors, acc, state)
+      {:ok, pid} when restart == :temporary ->
+        acc = [{pid, [ref | :undefined]} | acc]
+        start_children(extras, from, child, errors, acc, state)
+      {:ok, pid} ->
+        acc = [{pid, [ref | args]} | acc]
+        start_children(extras, from, child, errors, acc, state)
+      :ignore ->
+        start_children(extras, from, child, errors+1, acc, state)
+      {:error, reason} ->
+        :error_logger.error_msg('DynamicSupervisor failed to start child from: ~p with reason: ~p~n',
+          [from, reason])
+        report_error(:start_error, reason, :undefined, args, child, state)
+        start_children(extras, from , child, errors+1, acc, state)
+    end
+  end
+  defp start_children([], _, _, errors, acc, _) do
+    {acc, errors}
+  end
+
+  defp maybe_ask(ref, events, down, children, state) do
+    %{producers: producers} = state
+    case producers do
+      %{^ref => {to, count, demand, min, max}} when demand - events <= min ->
+        new_count = count + events - down
+        demand = demand - events
+        ask = (max - new_count) - demand
+        _ = if ask > 0, do: GenStage.ask(to, ask)
+        new_demand = demand + ask
+        new_entry = {to, new_count, new_demand, min, max}
+        producers = Map.put(producers, ref, new_entry)
+        %{state | children: children, producers: producers}
```

Comment:
I think we can merge this one with the clause above by simply removing the guard clause and they will behave the same.